### PR TITLE
Notification Merge to Dev

### PR DESCRIPTION
**Pull request by OneLeggedPigeon**

### CODE REVIEW

File: src/test/java/com/revature/studyforce/flashcard/integration/FlashcardIntegrationTest.java [hunk 1]
```diff
@@ -110,6 +110,41 @@ void getAllByDifficultyTest() throws Exception {
...
+                .andExpect(MockMvcResultMatchers.jsonPath("$.content[0].question").isString())
+                .andExpect(MockMvcResultMatchers.jsonPath("$.content[0].questionDifficultyTotal").isNumber())
+                .andExpect(MockMvcResultMatchers.jsonPath("$.content[0].questionDifficultyTotal").isNumber())
+                .andExpect(MockMvcResultMatchers.jsonPath("$.content[0].questionDifficultyAverage").isNumber())
+                .andExpect(MockMvcResultMatchers.jsonPath("$.content[0].createdTime").doesNotExist())
+                .andExpect(MockMvcResultMatchers.jsonPath("$.content[0].resolutionTime").doesNotExist());
+    }
+
+    @Test
+    void getAllByTopicTest() throws Exception {
```
Comment (EdsonR93) [inline src/test/java/com/revature/studyforce/flashcard/integration/FlashcardIntegrationTest.java:123]: better names for test cases

File: src/test/java/com/revature/studyforce/flashcard/integration/TopicIntegrationTest.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    private MockMvc mockMvc;
+
+    @Autowired
+    TopicController topicController;
+
+    @Autowired
+    TopicRepository topicRepository;
+
+    @Test
+    void topicIntegrationTest1() throws Exception {
```
Comment (EdsonR93) [inline src/test/java/com/revature/studyforce/flashcard/integration/TopicIntegrationTest.java:40]: better name for test case

File: src/test/java/com/revature/studyforce/flashcard/service/AnswerServiceTest.java [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+    }
+
+    @Test
+    void whenDeleteAnswerById(){
+        answerService.deleteAnswerById(1);
+        Mockito.verify(answerRepository, Mockito.times(1)).deleteById(1);
+    }
+
+    @Test
+    void whenGetAllByFlashcardId(){
```
Comment (EdsonR93) [inline src/test/java/com/revature/studyforce/flashcard/service/AnswerServiceTest.java:81]: better names

File: src/main/java/com/revature/studyforce/flashcard/dto/TopicDTO.java [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+
+import lombok.AllArgsConstructor;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+@NoArgsConstructor
+@AllArgsConstructor
+@Data
+public class TopicDTO {
+    String topic;
```
Comment (DanRey321) [inline src/main/java/com/revature/studyforce/flashcard/dto/TopicDTO.java:11]: Include documentation

File: src/main/java/com/revature/studyforce/notification/ServletInitializer.java [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+
+/**
+ * Used to configure HATOEOS links in JPA repository
+ * @author Ronald Lopez
+ */
+public class ServletInitializer extends SpringBootServletInitializer {
+    @Override
+    protected SpringApplicationBuilder configure(SpringApplicationBuilder application) {
+        return application.sources(StudyForceApplication.class);
+    }
```
Comment (nwickh03) [inline src/main/java/com/revature/studyforce/notification/ServletInitializer.java:15]: should be moved to package src/main/java/com/revature/studyforce/configuration
 or considered for deletion.

**Review by DanRey321 — CHANGES_REQUESTED** [review]
Several Files do not have any documentation. Some are missing @author and @links.
Controller files need @links to service layer methods used. Service methods need @link to repository methods used. Recommend every member go through files they created and include these requirements. Documentation Standard google doc available in Discord.

File: src/test/java/com/revature/studyforce/notification/integration/FlashcardSubscriptionIntegrationTest.java [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+
+    private MockMvc mockMvc;
+
+    @Autowired
+    private SubscriptionController subscriptionController;
+    @Autowired
+    private SubscriptionService subscriptionService;
+    @Autowired
+    private SubscriptionRepository subscriptionRepository;
+    @Autowired
```
Comment (EdsonR93) [inline src/test/java/com/revature/studyforce/notification/integration/FlashcardSubscriptionIntegrationTest.java:45]: change this to a constructor so you can make repos and services private and final

File: src/main/java/com/revature/studyforce/notification/config/RestConfig.java [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+            entityClasses.add(entityType.getJavaType());
+        }
+
+        // expose the entity ids for the array of the entity/domain types
+        Class[] domainTypes = entityClasses.toArray(new Class[0]);
+        config.exposeIdsFor(domainTypes);
+
+    }
+
+}
```
Comment (nwickh03) [inline src/main/java/com/revature/studyforce/notification/config/RestConfig.java:56]: should be moved to package src/main/java/com/revature/studyforce/configuration
or considered for deletion.

File: src/test/java/com/revature/studyforce/notification/service/FlashcardSubscriptionServiceTest.java [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+        flashcard = new Flashcard();
+        flashcard.setId(1);
+        subscription = new Subscription();
+        subscription.setId(1);
+        flashcardSubscription = new FlashcardSubscription(flashcard, subscription);
+        flashcardSubscriptions.add(flashcardSubscription);
+    }
+
+    @Test
+    void getFlashcardSubscriptionByFlashcardAndUserId(){
```
Comment (EdsonR93) [inline src/test/java/com/revature/studyforce/notification/service/FlashcardSubscriptionServiceTest.java:53]: test cases need better names

File: src/test/java/com/revature/studyforce/notification/service/StacktraceSubscriptionServiceTest.java [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+        stacktrace = new Stacktrace();
+        stacktrace.setStacktraceId(1);
+        subscription = new Subscription();
+        subscription.setId(1);
+        stacktraceSubscription = new StacktraceSubscription(stacktrace, subscription);
+        stacktraceSubscriptions.add(stacktraceSubscription);
+    }
+
+    @Test
+    void getStacktraceSubscriptionByStacktraceAndUserId(){
```
Comment (EdsonR93) [inline src/test/java/com/revature/studyforce/notification/service/StacktraceSubscriptionServiceTest.java:57]: better names for test cases

**Review by OneLeggedPigeon — COMMENTED** [review]
consider git checkout on these files:

git checkout origin/dev -- ____
pom.xml
src/test/resources/test-application.properties

git checkout origin/flashcard -- ____
src/main/java/com/revature/StudyForce/flashcard/*
src/test/java/com/revature/StudyForce/flashcard/*

git checkout origin/stacktrace -- ____
src/main/java/com/revature/StudyForce/stacktrace/*

File: src/test/java/com/revature/studyforce/notification/service/SubscriptionServiceTest.java [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+    public void setUp(){
+        user = new User();
+        user.setUserId(1);
+        subscription = new Subscription(1, user, "endpoint",  "p256dh", "auth");
+        subscriptions.add(subscription);
+        subscriptionPage = new PageImpl<>(subscriptions);
+    }
+
+    @Test
+    void getSubscriptionByUserIdTest(){
```
Comment (EdsonR93) [inline src/test/java/com/revature/studyforce/notification/service/SubscriptionServiceTest.java:49]: better names

**Review by EdsonR93 — CHANGES_REQUESTED** [review]
Some test classes require better names for test cases

File: src/main/java/com/revature/studyforce/notification/controller/SubscriptionController.java [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+
+/**
+ * Subscription Controller for
+ *  Subscriptions {@link SubscriptionService}
+ *  FlashcardSubscriptions {@link FlashcardSubscriptionService}
+ *  StacktraceSubscriptions {@link StacktraceSubscriptionService}
+ * @author Brandon Pinkerton
+ */
+@RestController
+@RequestMapping("/subscriptions")
```
Comment (nwickh03) [inline src/main/java/com/revature/studyforce/notification/controller/SubscriptionController.java:23]: '@ ResponseBody annotation is unneeded with @ RestController

File: src/test/java/com/revature/studyforce/notification/service/FlashcardSubscriptionServiceTest.java [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+
+@DirtiesContext(classMode = DirtiesContext.ClassMode.BEFORE_EACH_TEST_METHOD)
+@SpringBootTest
+@TestPropertySource(locations = "classpath:test-application.properties")
+public class FlashcardSubscriptionServiceTest {
```
Comment (DanRey321) [inline src/test/java/com/revature/studyforce/notification/service/FlashcardSubscriptionServiceTest.java:26]: @Author Documentation required and @link to service layer being tested.

File: src/test/java/com/revature/studyforce/notification/service/StacktraceSubscriptionServiceTest.java [hunk 13]
```diff
@@ -0,0 +1,143 @@
...
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+
+@DirtiesContext(classMode = DirtiesContext.ClassMode.BEFORE_EACH_TEST_METHOD)
+@SpringBootTest
+@TestPropertySource(locations = "classpath:test-application.properties")
+public class StacktraceSubscriptionServiceTest {
```
Comment (DanRey321) [inline src/test/java/com/revature/studyforce/notification/service/StacktraceSubscriptionServiceTest.java:30]: @Author Documentation required and @link to service layer being tested.

File: src/test/java/com/revature/studyforce/notification/service/SubscriptionServiceTest.java [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+import org.springframework.test.context.TestPropertySource;
+import static org.mockito.ArgumentMatchers.any;
+
+import java.util.ArrayList;
+import java.util.List;
+
+@DirtiesContext(classMode = DirtiesContext.ClassMode.BEFORE_EACH_TEST_METHOD)
+@SpringBootTest
+@TestPropertySource(locations = "classpath:test-application.properties")
+public class SubscriptionServiceTest {
```
Comment (DanRey321) [inline src/test/java/com/revature/studyforce/notification/service/SubscriptionServiceTest.java:26]: @Author Documentation required and @link to service layer it is testing.

**Review by DanRey321 — CHANGES_REQUESTED** [review]
Some testing files need proper documentation and better method names

File: src/main/java/com/revature/studyforce/notification/controller/SubscriptionController.java [hunk 15]
```diff
@@ -0,0 +1,155 @@
...
+    @DeleteMapping("/stacktraces")
+    public @ResponseBody
+    StacktraceSubscription deleteStacktraceSubscription(@RequestBody StacktraceSubscriptionDTO subscriptionRequest){
+        return stacktraceSubscriptionService.deleteStacktraceSubscription(
+                subscriptionRequest.getStacktraceId(),
+                subscriptionRequest.getUserId()
+        );
+    }
+
+}
```
Comment (nwickh03) [inline src/main/java/com/revature/studyforce/notification/controller/SubscriptionController.java:155]: fix linting issues
mapping needs tweaked to meet rest standards

File: src/main/java/com/revature/studyforce/notification/repository/FlashcardSubscriptionRepository.java [hunk 16]
```diff
@@ -0,0 +1,20 @@
...
+/**
+ * Reposiotry for Flashcard Subscriptiosn {@link FlashcardSubscription}
+ * @author Brandon Pinkerton
+ */
+@Repository
+public interface FlashcardSubscriptionRepository extends JpaRepository<FlashcardSubscription, Integer> {
+    FlashcardSubscription findByFlashcardSubscriptionId(FlashcardSubscriptionId flashcardSubscriptionId);
+    List<FlashcardSubscription> findAllByFlashcard_Id(Integer flashcardId);
+    List<FlashcardSubscription> findAllBySubscription_User_UserId(Integer userId);
+}
```
Comment (nwickh03) [inline src/main/java/com/revature/studyforce/notification/repository/FlashcardSubscriptionRepository.java:20]: Refactor with pagination

File: src/main/java/com/revature/studyforce/notification/repository/StacktraceSubscriptionRepository.java [hunk 17]
```diff
@@ -0,0 +1,37 @@
...
+     */
+    List<StacktraceSubscription> findAllByStacktrace_StacktraceId(Integer stacktraceId);
+
+    /**
+     * Find all the subscriptions for a single user
+     * @param userId
+     * @return
+     */
+    List<StacktraceSubscription> findAllBySubscription_User_UserId(Integer userId);
+}
```
Comment (nwickh03) [inline src/main/java/com/revature/studyforce/notification/repository/StacktraceSubscriptionRepository.java:37]: Refactor with pagination.

File: src/main/java/com/revature/studyforce/notification/service/FlashcardSubscriptionService.java [hunk 18]
```diff
@@ -0,0 +1,130 @@
...
+
+    /**
+     * Retrieves all flashcard subscriptions for a given user {@link FlashcardSubscriptionRepository#findAllBySubscription_User_UserId(Integer)}
+     * @param userId The userId of the user to retrieve all of their subscriptions
+     * @return A list of all flashcard subscriptions to a specific user
+     */
+    public List<FlashcardSubscription> getAllSubscriptionsByUserId(Integer userId){
+        return flashcardSubscriptionRepository.findAllBySubscription_User_UserId(userId);
+    }
+
```
Comment (nwickh03) [inline src/main/java/com/revature/studyforce/notification/service/FlashcardSubscriptionService.java:66]: Refactor with pagination

File: src/test/java/com/revature/studyforce/notification/integration/FlashcardSubscriptionIntegrationTest.java [hunk 19]
```diff
@@ -0,0 +1,121 @@
...
+import org.springframework.test.web.servlet.setup.MockMvcBuilders;
+
+import java.sql.Timestamp;
+import java.time.LocalDateTime;
+
+@DirtiesContext(classMode = DirtiesContext.ClassMode.BEFORE_EACH_TEST_METHOD)
+@SpringBootTest
+@AutoConfigureMockMvc
+@TestPropertySource(locations = "classpath:test-application.properties")
+public class FlashcardSubscriptionIntegrationTest {
```
Comment (DanRey321) [inline src/test/java/com/revature/studyforce/notification/integration/FlashcardSubscriptionIntegrationTest.java:35]: @author Documentation required and @link to service layer it is testing.

File: src/test/java/com/revature/studyforce/notification/integration/StacktraceSubscriptionIntegrationTest.java [hunk 20]
```diff
@@ -0,0 +1,134 @@
...
+import org.springframework.test.web.servlet.setup.MockMvcBuilders;
+
+import java.sql.Timestamp;
+import java.time.LocalDateTime;
+
+@DirtiesContext(classMode = DirtiesContext.ClassMode.BEFORE_EACH_TEST_METHOD)
+@SpringBootTest
+@AutoConfigureMockMvc
+@TestPropertySource(locations = "classpath:test-application.properties")
+public class StacktraceSubscriptionIntegrationTest {
```
Comment (DanRey321) [inline src/test/java/com/revature/studyforce/notification/integration/StacktraceSubscriptionIntegrationTest.java:40]: @author Documentation required and @link to service layer it is testing.

**Review by DanRey321 — CHANGES_REQUESTED** [review]
Some testing files need proper documentation and better method names

File: src/main/java/com/revature/studyforce/notification/service/FlashcardSubscriptionService.java [hunk 21]
```diff
@@ -0,0 +1,130 @@
...
+    /**
+     * Generates a list of subscriptions for a given user to be deleted by /deleteAllFlashcardSubscriptions
+     * @param userId The userId of the user to delete all subscriptions.
+     * @return The list of deleted subscriptions.
+     */
+    public List<FlashcardSubscription> deleteAllFlashcardSubscriptionsByUserId(Integer userId){
+        List<FlashcardSubscription> subscriptions = getAllSubscriptionsByUserId(userId);
+        return deleteAllFlashcardSubscriptions(subscriptions);
+    }
+}
```
Comment (nwickh03) [inline src/main/java/com/revature/studyforce/notification/service/FlashcardSubscriptionService.java:130]: Refactor with pagination

File: src/main/java/com/revature/studyforce/notification/service/StacktraceSubscriptionService.java [hunk 22]
```diff
@@ -0,0 +1,136 @@
...
+    }
+
+    /**
+     * Retrieves all stacktraces subscriptions for a given user {@link StacktraceSubscriptionRepository#findAllBySubscription_User_UserId(Integer)}
+     * @param userId The userId of the user to retrieve all of their subscriptions
+     * @return A list of all stacktraces subscriptions to a specific user
+     */
+    public List<StacktraceSubscription> getAllSubscriptionsByUserId(Integer userId){
+        return stacktraceSubscriptionRepository.findAllBySubscription_User_UserId(userId);
+    }
```
Comment (nwickh03) [inline src/main/java/com/revature/studyforce/notification/service/StacktraceSubscriptionService.java:71]: Refactor with pagination

File: src/main/java/com/revature/studyforce/notification/service/StacktraceSubscriptionService.java [hunk 23]
```diff
@@ -0,0 +1,136 @@
...
+    /**
+     * Generates a list of subscriptions for a given user to be deleted by /deleteAllStacktraceSubscriptions
+     * @param userId The userId of the user to delete all subscriptions.
+     * @return The list of deleted subscriptions.
+     */
+    public List<StacktraceSubscription> deleteAllStacktraceSubscriptionsByUserId(Integer userId){
+        List<StacktraceSubscription> subscriptions = getAllSubscriptionsByUserId(userId);
+        return deleteAllStacktraceSubscriptions(subscriptions);
+    }
+}
```
Comment (nwickh03) [inline src/main/java/com/revature/studyforce/notification/service/StacktraceSubscriptionService.java:136]: Refactor with pagination
Check linting

**Review by nwickh03 — CHANGES_REQUESTED** [review]

File: src/main/java/com/revature/studyforce/notification/repository/FlashcardSubscriptionRepository.java (hunk 16 above)
Comment (bpinkerton) [reply src/main/java/com/revature/studyforce/notification/repository/FlashcardSubscriptionRepository.java:20]: This is an internal service method, not something to be hit by an endpoint and served in the frontend.

File: src/main/java/com/revature/studyforce/notification/repository/StacktraceSubscriptionRepository.java (hunk 17 above)
Comment (bpinkerton) [reply src/main/java/com/revature/studyforce/notification/repository/StacktraceSubscriptionRepository.java:37]: This is an internal service method, not something to be hit by an endpoint and served in the frontend.

File: src/main/java/com/revature/studyforce/notification/service/FlashcardSubscriptionService.java (hunk 18 above)
Comment (bpinkerton) [reply src/main/java/com/revature/studyforce/notification/service/FlashcardSubscriptionService.java:66]: This is an internal service method, not something to be hit by an endpoint and served in the frontend.

File: src/main/java/com/revature/studyforce/notification/service/FlashcardSubscriptionService.java (hunk 21 above)
Comment (bpinkerton) [reply src/main/java/com/revature/studyforce/notification/service/FlashcardSubscriptionService.java:130]: This is an internal service method, not something to be hit by an endpoint and served in the frontend.

File: src/main/java/com/revature/studyforce/notification/service/StacktraceSubscriptionService.java (hunk 22 above)
Comment (bpinkerton) [reply src/main/java/com/revature/studyforce/notification/service/StacktraceSubscriptionService.java:71]: This is an internal service method, not something to be hit by an endpoint and served in the frontend.

File: src/main/java/com/revature/studyforce/notification/service/StacktraceSubscriptionService.java (hunk 23 above)
Comment (bpinkerton) [reply src/main/java/com/revature/studyforce/notification/service/StacktraceSubscriptionService.java:136]: This is an internal service method, not something to be hit by an endpoint and served in the frontend.

File: src/main/java/com/revature/studyforce/notification/controller/SubscriptionController.java (hunk 15 above)
Comment (bpinkerton) [reply src/main/java/com/revature/studyforce/notification/controller/SubscriptionController.java:155]: Looking into it

File: src/main/java/com/revature/studyforce/notification/controller/SubscriptionController.java (hunk 11 above)
Comment (bpinkerton) [reply src/main/java/com/revature/studyforce/notification/controller/SubscriptionController.java:23]: fixed

File: src/test/java/com/revature/studyforce/notification/integration/StacktraceSubscriptionIntegrationTest.java (hunk 20 above)
Comment (bpinkerton) [reply src/test/java/com/revature/studyforce/notification/integration/StacktraceSubscriptionIntegrationTest.java:40]: fixed

File: src/test/java/com/revature/studyforce/notification/integration/FlashcardSubscriptionIntegrationTest.java (hunk 19 above)
Comment (bpinkerton) [reply src/test/java/com/revature/studyforce/notification/integration/FlashcardSubscriptionIntegrationTest.java:35]: fixed

File: src/test/java/com/revature/studyforce/notification/service/SubscriptionServiceTest.java (hunk 14 above)
Comment (bpinkerton) [reply src/test/java/com/revature/studyforce/notification/service/SubscriptionServiceTest.java:26]: fixed

File: src/test/java/com/revature/studyforce/notification/service/StacktraceSubscriptionServiceTest.java (hunk 13 above)
Comment (bpinkerton) [reply src/test/java/com/revature/studyforce/notification/service/StacktraceSubscriptionServiceTest.java:30]: fixed

File: src/test/java/com/revature/studyforce/notification/service/FlashcardSubscriptionServiceTest.java (hunk 12 above)
Comment (bpinkerton) [reply src/test/java/com/revature/studyforce/notification/service/FlashcardSubscriptionServiceTest.java:26]: fixed

File: src/test/java/com/revature/studyforce/flashcard/integration/FlashcardIntegrationTest.java (hunk 1 above)
Comment (bpinkerton) [reply src/test/java/com/revature/studyforce/flashcard/integration/FlashcardIntegrationTest.java:123]: not our area

File: src/test/java/com/revature/studyforce/flashcard/integration/TopicIntegrationTest.java (hunk 2 above)
Comment (bpinkerton) [reply src/test/java/com/revature/studyforce/flashcard/integration/TopicIntegrationTest.java:40]: not our area

File: src/test/java/com/revature/studyforce/flashcard/service/AnswerServiceTest.java (hunk 3 above)
Comment (bpinkerton) [reply src/test/java/com/revature/studyforce/flashcard/service/AnswerServiceTest.java:81]: not our area

File: src/test/java/com/revature/studyforce/notification/integration/FlashcardSubscriptionIntegrationTest.java (hunk 6 above)
Comment (bpinkerton) [reply src/test/java/com/revature/studyforce/notification/integration/FlashcardSubscriptionIntegrationTest.java:45]: omitted

File: src/test/java/com/revature/studyforce/notification/service/FlashcardSubscriptionServiceTest.java (hunk 8 above)
Comment (bpinkerton) [reply src/test/java/com/revature/studyforce/notification/service/FlashcardSubscriptionServiceTest.java:53]: fixed

File: src/test/java/com/revature/studyforce/notification/service/SubscriptionServiceTest.java (hunk 10 above)
Comment (bpinkerton) [reply src/test/java/com/revature/studyforce/notification/service/SubscriptionServiceTest.java:49]: fixed

File: src/test/java/com/revature/studyforce/notification/service/StacktraceSubscriptionServiceTest.java (hunk 9 above)
Comment (bpinkerton) [reply src/test/java/com/revature/studyforce/notification/service/StacktraceSubscriptionServiceTest.java:57]: fixed

File: src/main/java/com/revature/studyforce/notification/ServletInitializer.java (hunk 5 above)
Comment (bpinkerton) [reply src/main/java/com/revature/studyforce/notification/ServletInitializer.java:15]: deleted

File: src/main/java/com/revature/studyforce/notification/config/RestConfig.java (hunk 7 above)
Comment (bpinkerton) [reply src/main/java/com/revature/studyforce/notification/config/RestConfig.java:56]: deleted

**Review by EdsonR93 — APPROVED** [review]

**Review by DanRey321 — APPROVED** [review]
Good

File: src/main/java/com/revature/studyforce/notification/repository/StacktraceSubscriptionRepository.java (hunk 17 above)
Comment (nwickh03) [inline src/main/java/com/revature/studyforce/notification/repository/StacktraceSubscriptionRepository.java:37]: Google linting is giving warnings about empty JavaDoc tags x6

**Review by nwickh03 — CHANGES_REQUESTED** [review]
Looks mostly good. One linting/documentation issue.

**Review by nwickh03 — APPROVED** [review]

**Pull request merged**